File: backend/app/models/reading_progress.py
from sqlalchemy import Column, Integer, Float, DateTime, ForeignKey, UniqueConstraint
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database.postgres import Base


class ReadingProgress(Base):
    __tablename__ = "reading_progress"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    post_id = Column(Integer, ForeignKey("posts.id"), nullable=False)
    current_page = Column(Integer, nullable=False, default=1)
    total_pages = Column(Integer, nullable=False, default=1)
    progress_percentage = Column(Float, nullable=False, default=0.0)
    last_read_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    reading_time_minutes = Column(Integer, default=0)  # Total time spent reading
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    # Unique constraint: one progress record per user per post
    __table_args__ = (UniqueConstraint('user_id', 'post_id', name='_user_post_progress_uc'),)
    
    # Relationships
    user = relationship("User", backref="reading_progress")
    post = relationship("Post", backref="reading_progress")

File: frontend/src/api/readingProgress.ts
import axiosClient from './axiosClient'

export interface ReadingProgress {
  id: number
  user_id: number
  post_id: number
  current_page: number
  total_pages: number
  progress_percentage: number
  reading_time_minutes: number
  last_read_at: string
  created_at: string
  updated_at?: string
}

export interface ReadingProgressUpdate {
  current_page?: number
  total_pages?: number
  progress_percentage?: number
  reading_time_minutes?: number
}

export interface ReadingStats {
  total_books_read: number
  total_reading_time_minutes: number
  total_pages_read: number
  average_completion: number
}

export const readingProgressApi = {
  getProgress: async (postId: number): Promise<ReadingProgress> => {
    const response = await axiosClient.get(`/api/v1/reading-progress/post/${postId}`)
    return response.data
  },

  updateProgress: async (postId: number, data: ReadingProgressUpdate): Promise<ReadingProgress> => {
    const response = await axiosClient.put(`/api/v1/reading-progress/post/${postId}`, data)
    return response.data
  },

  getStats: async (): Promise<ReadingStats> => {
    const response = await axiosClient.get('/api/v1/reading-progress/stats')
    return response.data
  },
}

